plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergConfig.java

```diff
@@ -497,4 +498,17 @@ public IcebergConfig setMetadataCacheEnabled(boolean metadataCacheEnabled)
         this.metadataCacheEnabled = metadataCacheEnabled;
         return this;
     }
+
+    public boolean isBucketExecutionEnabled()
+    {
+        return bucketExecutionEnabled;
+    }
+
+    @Config("iceberg.bucket-execution")
+    @ConfigDescription("Enable bucket-aware execution: use physical bucketing information to optimize queries")
+    public IcebergConfig setBucketExecutionEnabled(boolean bucketExecutionEnabled)
```

Comment:
note that it won't work anyway because `DetermineTableScanNodePartitioning` uses:
```
        int numberOfBuckets = bucketNodeMap.map(ConnectorBucketNodeMap::getBucketCount)
                .orElseGet(() -> nodePartitioningManager.getNodeCount(context.getSession(), partitioning.partitioningHandle()));
```

It's essential for connector to be able to provide bucket count to the engine.

If it's just scanning one partition, then it doesn't make sense to funnel that though single node (unless the partition is super small itself).

Reply:
The iceberg and hive code both only handle the pushdown if hash bucketing is used.  This is not used for value based.

Reply:
I think you miss the point. CBO (`DetermineTableScanNodePartitioning`) decides whether to use connector provided bucketing or not. `DetermineTableScanNodePartitioning` won't use connector bucketing if it doesn't know how many "buckets" there are

Reply:
Yes, but in AddExchanges we perform pushdown and if it works, we then directly call DetermineTableScanNodePartitioning so it can decide if we should actually use the bucketing from the connector.  So both systems must decide that the partitioning should be used for execution.

Reply:
@dain `DetermineTableScanNodePartitioning` will never choose to use Iceberg partitioning in current code, because it doesn't know partition count.

Reply:
I assure it it does.  When you have have a hash bucket function it uses that number. You can pull the branch and run it yourself.

Reply:
> When you have have a hash bucket function it uses that number

What is the number? Is it artificial or it actually represents how many partitions there are? What happens if there is only single partition scanned?

The code in `io.trino.sql.planner.iterative.rule.DetermineTableScanNodePartitioning#apply` is:
```
        int numberOfBuckets = bucketNodeMap.map(ConnectorBucketNodeMap::getBucketCount)
                .orElseGet(() -> nodePartitioningManager.getNodeCount(context.getSession(), partitioning.getPartitioningHandle()));
        int numberOfTasks = max(taskCountEstimator.estimateSourceDistributedTaskCount(context.getSession()), 1);

        return Result.ofPlanNode(node
                .withUseConnectorNodePartitioning((double) numberOfBuckets / numberOfTasks >= getTableScanNodePartitioningMinBucketToTaskRatio(context.getSession())));
```

If the number of buckets is some arbitrary big number, but only single partition is scanned, then such query won't utilize the cluster properly (it will essentially utilize single node).

Reply:
When you have partitioning that looks like:
```
partitioning = ARRAY['bucket(key1, 13)', 'bucket(key2, 17)']
```
If you group on `key1`, you get 13 buckets, and 17 for `key2`.

The answer to the question about number of buckets actually scanned is not affected by my code at all, so whatever happened before still happens.  I believe (but don't know), you would likely end up with a constant for the partition funcation argument, and there is logic that disables the partitioning in that case.  Regardless, this code actually improves this in more cases, because partitioning is only enabled, when the query needs it.  Otherwise the table is read as if it were not partitioned.

Reply:
@dain 
Does your PR also enable `DetermineTableScanNodePartitioning` for Iceberg `partitioning = ARRAY['c1', 'c2'].`?
If so, then it's a new behavior which would lead to lack of query parallelism for single or two partitions scans.